cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceRuntimeService.java

```diff
@@ -567,11 +595,14 @@ private ProgramState getProgramState(boolean success, String failureInfo) {
    * Calls the destroy method of {@link ProgramLifecycle}.
    */
   private void destroy(final boolean succeeded, final String failureInfo, final ClassLoader mapReduceClassLoader) {
-    TransactionContext txContext = context.getTransactionContext();
+    final TransactionControl txControl = mapReduce instanceof ProgramLifecycle
+      ? Transactions.getTransactionControl(TransactionControl.IMPLICIT, MapReduce.class, mapReduce, "destroy")
+      : TransactionControl.IMPLICIT;
     try {
-      Transactions.execute(txContext, "destroy", new Callable<Void>() {
+      TransactionContext txContext = context.getTransactionContext();
+      boolean success = Transactions.execute(txContext, "destroy", new Callable<Boolean>() {
```

Comment:
Similar to above, if the tx control is `EXPLICIT`, we are creating a tx here just to execute the `context.setState` call, which actually doesn't need to be transactional. I think it's cleaner if we just have a if-else logic for the implicit and explicit case.